### PR TITLE
[#15] Investigate whether we can evaluate code lazily if needed

### LINKED ISSUE #15: Investigate whether we can load code lazily if needed
# Clarification and motivation

The contract we are going to implement is not quite small and we are very likely to hit some limits that Tezos has, namely gas limit and operation size limit. There are at least three ways to resolve such a problem if it appears:
1. Make some local optimizations.
2. Throw away some code (e. g. unnecessary getters).
3. Load code lazily.

The first two options are quite limited. We may be able to throw away something, but we shouldn't rely on that in any case (obviously). So chances are that we will have to use the third option.

Probably the best way to do that is to have such storage:
```
data Storage =
  { entrypoint1 :: BigMap () (Lambda (param1, Storage') ([Operation], Storage'))
  , entrypoint2 :: BigMap () (Lambda (param2, Storage') ([Operation], Storage'))
  ...
  , storage :: Storage'  -- actual data
  }
```

In this case all code remains typed, but is stored in `big_map`s and is loaded lazily, only when we use it. In practice, each contract call will load only one entrypoint and it should save a lot of gas. Also it should decrease origination operation size because we can fill these `big_map`s after origination.

Currently we need to figure out whether this approach is possible to implement in LIGO.

# Acceptance criteria

Implement a contract that stores one integer and takes a `or (pair int int) int`. In the first case it puts the sum of pair's items to the storage, in the latter case it puts the `int` itself. It should be implemented using the aforementioned pattern. In LIGO of course. Just a proof of concept.

**OR**

If you find it impossible, explain why.

### PR DESCRIPTION
**Pull request by awkure**
## Related issue(s)

Resolves #15 

#### Related changes (conditional)

- Tests
  - [X] If I added new functionality, I added tests covering it.
  - [X] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [X] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [X] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [X] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by awkure** [comment]
With current implementation the execution of seemingly correct commands fail with following error log
```
$ ligo run-function test.ligo main "(Sum (big_map [(unit) -> (sum, (1, 2))]), 2)"                                                     [3ebc4e8]
ligo: error parsing program code
$ ligo compile-contract test.ligo main                                                                                              
ligo: not a comparable type: unit
```

Note that the contract itself parses correctly (for instance if we define dummy `unit -> unit` lambda and execute it from cli).

**Comment by gromakovsky** [comment]
> ligo: not a comparable type: unit

Well, it makes sense because `unit` is not comparable indeed. Let's replace it with `bool` then :man_shrugging: 

So please do it and then share generated Michelson contract.

**Comment by awkure** [comment]
But semantically it should be because we have to have a unit for comparable semigroup in any case

File: test.tz [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{ parameter
+    (big_map
```
Comment (gromakovsky) [inline test.tz:2]: No, there should be 1 big_map per entrypoint. So in this case there should be 2 big_maps.

**Comment by awkure** [comment]
Somewhat strange seeing the following error when trying to execute it 
```
$ ligo run-function test.ligo main "(Sum (1,2), record [sum_entry = big_map [(True) -> sum]; put_entry = big_map [(True) -> put]; state=0])" 
ligo: generated application. different kinds:  {"a":"int","b":"record[put_entry -> (TO_Big Map (bool,( int * int ) -> ( (TO_list(operation)) * int ))) , state -> int , sum_entry -> (TO_Big Map (bool,( ( int * int ) * int ) -> ( (TO_list(operation)) * int )))]"}
```

Meaning that the ligo compiller somewhat skips the second argument in passed record and somewhat substitutes expected `state` with passed `put_entry`.

**Comment by gromakovsky** [comment]
Is it possible to print lambdas that should be put into storage?

**Comment by gromakovsky** [comment]
As a backup plan we can even write this template in Lorentz and implement business logic in LIGO. Then we will print lambdas from LIGO code and put them into the storage of the template contract written in Lorentz.

**Comment by awkure** [comment]
Uhm, no. There's no any printing capability currently available in LIGO. It prints values only when it returns some.

**Comment by gromakovsky** [comment]
See https://github.com/tqtezos/stablecoin/issues/15#issuecomment-635437548

**Pull request closed without merging**